src/coreclr/vm/comsynchronizable.cpp

```diff
@@ -899,3 +892,49 @@ extern "C" void QCALLTYPE ThreadNative_ResetAbort()
         pThread->UnmarkThreadForAbort(EEPolicy::TA_Safe);
     }
 }
+
+extern "C" int32_t QCALLTYPE SyncTable_AssignEntry(QCall::ObjectHandleOnStack obj)
+{
+    QCALL_CONTRACT;
+
+    int32_t index = -1;
+
+    BEGIN_QCALL;
+
+    GCX_COOP();
+
+    // Force creation of a SyncBlock for the object.
+    index = (int32_t)obj.Get()->GetSyncBlock()->GetSyncBlockIndex();
+
+    END_QCALL;
+
+    return index;
+}
+
+FCIMPL1(OBJECTHANDLE, SyncTable_GetLockHandleIfExists, int idx)
+{
+    FCALL_CONTRACT;
+
+    _ASSERTE(0 <= idx && idx <= SyncBlockCache::GetSyncBlockCache()->GetTableEntryCount());
+    return SyncTableEntry::GetSyncTableEntry()[idx].m_SyncBlock->GetLockIfExists();
+}
+FCIMPLEND
+
+extern "C" void QCALLTYPE SyncTable_GetLockObject(int idx, QCall::ObjectHandleOnStack obj)
+{
+    QCALL_CONTRACT;
+
+    BEGIN_QCALL;
+
+    _ASSERTE(0 <= idx && idx <= SyncBlockCache::GetSyncBlockCache()->GetTableEntryCount());
+
+    OBJECTHANDLE lock = SyncTableEntry::GetSyncTableEntry()[idx].m_SyncBlock->GetLock();
```

Comment:
It is not safe to access `g_pSyncTable`  in cooperative mode. `g_pSyncTable` can grow and reclamation of the previous block is synchronized using GC suspension.

If you access `g_pSyncTable` in pre-emptive mode, you are at risk of accessing memory that was freed.

Reply:
I'll move the sync block retrieval from the table back into coop mode.